src/OrleansProviders/Streams/Common/PooledCache/PooledQueueCache.cs

```diff
@@ -0,0 +1,306 @@
+                    }
+                    else
+                    {
+                        cursor.CurrentBlock.Value.TryFindNextMessage(cursor.Index + 1, cursor.StreamGuid, cursor.StreamNamespace, out index);
+                    }
+                    cursor.Index = index;
+                }
+
+                // check if this message is in the cursor's stream
+                if (cacheDataAdapter.IsInStream(ref currentMessage, cursor.StreamGuid, cursor.StreamNamespace))
+                {
+                    message = cacheDataAdapter.GetBatchContainer(ref currentMessage);
+                    cursor.SequenceToken = cursor.CurrentBlock.Value.GetSequenceToken(cursor.Index);
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        public void Add(TQueueMessage message)
+        {
+            if (message == null)
+            {
+                throw new ArgumentNullException("message");
+            }
+
+            PerformPendingPurges();
+
+            // allocate message from pool
+            CachedMessageBlock<TQueueMessage, TCachedMessage> block = pool.AllocateMessage(message);
+
+            // If new block, add message block to linked list
+            if (block != messageBlocks.FirstOrDefault())
+                messageBlocks.AddFirst(block.Node);
+
+            PerformPendingPurges();
+        }
+
+        /// <summary>
+        /// Record that we need to purge all messages associated with purge request in the next purge cycle
+        /// </summary>
+        /// <param name="purgeRequest"></param>
+        public void Purge(IDisposable purgeRequest)
+        {
+            purgeQueue.Enqueue(purgeRequest);
+        }
+
+        private void PerformPendingPurges()
+        {
+            uint purgeCount = 0;
+
+            // If there are purge requests, perform purges
+            if (!purgeQueue.IsEmpty)
+            {
+                IDisposable purgeRequest;
+                while (purgeQueue.TryDequeue(out purgeRequest))
+                {
+                    purgeCount += PerformBlockPurge(purgeRequest);
+                }
+            }
+        }
+
+        private uint PerformBlockPurge(IDisposable purgeRequest)
+        {
+            uint purgeCount = 0;
+            while (!IsEmpty &&
+                   cacheDataAdapter.ShouldPurge(messageBlocks.Last.Value[messageBlocks.Last.Value.OldestMessageIndex], purgeRequest)) // value has the same resource blockId as the one we are purging
+            {
+                messageBlocks.Last.Value.Remove();
+                // if block is currently empty, but all capacity has been exausted, remove
+                if (messageBlocks.Last.Value.IsEmpty && !messageBlocks.Last.Value.HasCapacity)
+                {
+                    messageBlocks.Last.Value.Dispose();
+                    messageBlocks.RemoveLast();
```

Comment:
I understand this is probably not called from multiple threads, but still for readability it might be good to store `messageBlocks.Last.Value` in a variable and then do `Remove()`, `IsEmpty`,  `HasCapacity` and `Dispose()` on it